keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,220 @@
+
+In addition to supporting cloud provider migration this KEP supports allowing users to support multiple container registries across different cloud providers by introducing a pluggable interface to cloud specific credential providers such that a single user could run multiple credential providers for different clouds at the same time. Currently this functionality is gated by having only a single, active cloud provider within the API server and kubelet.
+
+### Goals
+
+* Develop/test/release an interface for kubelet to obtain registry credentials from a cloud provider specific binary
+* Update/test/release the credential acquisition logic within kubelet
+* Build user documentation for out-of-tree credential providers
+* Migrate existing in-tree credential providers to new credential provider interface
+* Remove in-tree credential provider code from kuberentes core
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be installed on each node. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This Architecture is similar to the approach taken by CNI and is a well understood deployment pattern.  The API types are modelled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider API type will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URIs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Spec holds configuration for the credential provider
+    Spec RegistryCredentialProviderSpec `json:"spec,omitempty"`
+}
+
+type RegistryCredentialProviderSpec struct {
+    // ImageMatchRegex is used to match against the Image URI to determine which images to
+    // provide credentials for
+    ImageMatchRegex string
+
+    // Exec specifies a custom exec-based authentication plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
```

Comment:
Should there be an option to pass the image URL to the plugin?

Reply:
Yes, this would be part of the JSON encoded stdin input to the plugin.